### PR TITLE
Retarget Pulls Based on Pulls on Merge

### LINKED ISSUE #18408: Allow cascade PRs
### Feature Description

Sometime you need to have 'cascade' PRs. i.e.

* first you're working on feature A (derived from main branch) and you create PR feature A -> main
* next you need to work on feature B, but this feature depends on changes of feature A and it's corresponding PR is not yet approved. The best way to work on such feature is derive from feature A and later you need to create PR feature B -> feature A.

Currently second PR will be automatically closed on feature A merge and branch delete. It would be helpfull to rebase second PR. i.e. change 'feature B -> feature A' to 'feature B -> main' when first 'feature A -> main' is merged and 'feature A' branch is deleted.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by kvaster**
Sometimes you need to work on feature which depends on another feature which is in a review process. In this case you may create PR based on that feature insteadof main branch. Currently such PR will be closed without possibility to reopen in case parent feature will be approved, merged and it's branch will be deleted. Automatic target branch change will make life a lot easier in such cases. Github and Bitbucket behaves in such way.

Example:
PR1: main <- feature 1
PR2: feature <- feature2

Currently merging PR1 and deleteing it's branch will lead to PR2 closed without possibility to reopen without loosing it's review histort.

With this change PR2 will change it's target branch to main (PR2: main <- feature2) after PR1 will be merged and it's branch will be deleted.

fixes #18408

### CODE REVIEW

**Comment by lunny** [comment]
The PRs should be rebased when base branch updated but not when merging or deleting the branch.

**Comment by kvaster** [comment]
I think I've used wrong words. This PR is about changing branches target. I will change names.

**Comment by kvaster** [comment]
I've updated descritpion and names inside PR. Hopefully it's more clear now...

**Comment by spiffytortoise** [comment]
Chiming in since a stale label got added.

Having PRs get their base branch updated when a PR for the base branch is merged like this would be great. My team previously used Bitbucket's equivalent feature often for landing complex features in multiple stages.

**Comment by lunny** [comment]
I think Gitea should not change the target branch automatically but it could be changed manually.

**Comment by kvaster** [comment]
Currently you're loosing all PRs discussions in such case. Can it be optional setting for gitea behaviour ?

**Comment by spiffytortoise** [comment]
How would the workflow for manually changing the target branch look? I can think of two ways currently that both have problems:

Example PRs:

main <- feature1
feature1 <- feature2

#### Manual Scenario 1, merge oldest to newest:

1. Merge feature1 into main. Don't delete the feature1 branch, or else the feature2 PR will close.
2. Change the base branch of feature2's PR from feature1 to main.
3. Delete the feature1 branch.
4. Merge feature2 into main.

It's too easy to delete feature1 at the wrong time and permanently close the feature2 PR by accident. (#18478)

#### Manual Scenario 2, merge newest to oldest:

1. Merge feature2 into feature1.
2. feature1's PR ends up with feature2's changes in its diff. 
3. Merge feature1 into main.

The diff change in step 2 adds a bunch of noise to the PR, which makes it more difficult to understand the PR after.

#### Auto-Update Scenario, merge oldest to newest:

In comparison, this is how Bitbucket Server works:

1. Merge feature1 into main. Check "Delete feature1 and update the base branch of PRs targeting feature1 to main".
2. Bitbucket atomically updates the base branch of feature2 from feature1 to main, without any additional diff being added or displayed in feature2's PR.
3. Merge feature2 into main.

Why should Gitea not update PR base branches automatically?

**Comment by kvaster** [comment]
I've updated PR according to upstream.

Any suggestion in what section to add flag for enabling/disabling this feature ?

**Comment by kvaster** [comment]
One more PR update - make an option for retarget.

**Comment by 6543** [comment]
https://github.com/go-gitea/gitea/pull/18478#issuecomment-1112471939

hmm I think we can add some nice integration tests for mentioned cases

example:
https://github.com/go-gitea/gitea/blob/9efa47131f3fa576bd0ef73fa4c5b96c95d89906/integrations/git_test.go#L487-L498
....

File: modules/setting/repository.go [hunk 1]
```diff
@@ -78,6 +78,7 @@ var (
 			DefaultMergeMessageOfficialApproversOnly bool
 			PopulateSquashCommentWithCommitMessages  bool
 			AddCoCommitterTrailers                   bool
+			RetargetChildsOnClose                    bool
```
Comment (delvh) [inline modules/setting/repository.go]:
```suggestion
			RetargetChildrenOnClose                    bool
```

File: modules/setting/repository.go [hunk 2]
```diff
@@ -199,6 +200,7 @@ var (
...
```
Comment (delvh) [inline modules/setting/repository.go]:
```suggestion
			RetargetChildrenOnClose                    bool
```
Also when you add configuration options, please also add them to `custom/conf/app.example.ini` and `docs/content/doc/advanced/config-cheat-sheet.en-us.md`.

**Comment by lafriks** [comment]
Does option is really needed for this? I mean is there a use case where re-targeting would be undesirable?

**Comment by lunny** [comment]
> I personaly would not mind if it always try retargeting ... cc @lunny

Maybe manually allowing to retarget pulls base branch even if base branch(or base repository) was deleted is better or it's the first step of this auto retarget.

**Comment by kvaster** [comment]
Will try to do this myself today and will get back tomorrow if will need any help.

**Comment by night-gold** [comment]
Hello, is this still an ongoing PR or has it been abandoned?

File: routers/web/repo/pull.go [hunk 3]
```diff
@@ -1318,6 +1318,15 @@ func CleanUpPullRequest(ctx *context.Context) {
...
+
+	if setting.Repository.PullRequest.RetargetChildsOnClose && pr.BaseRepoID == pr.HeadRepoID {
+		if err := pull_service.RebaseBranchPulls(ctx, ctx.Doer, pr.HeadRepoID, pr.HeadBranch, pr.BaseBranch); err != nil {
+			log.Error("RebaseBranchPulls: %v", err)
+			ctx.Flash.Error(ctx.Tr("repo.branch.deletion_failed", fullBranchName))
+			return
+		}
+	}
+
 	if err := repo_service.DeleteBranch(ctx.Doer, pr.HeadRepo, gitRepo, pr.HeadBranch); err != nil {
```
Comment (delvh) [inline routers/web/repo/pull.go]: Why define this block twice when you can also move it into `repo_service.DeleteBranch` shown at the bottom and enhance the error handling even further below?

File: modules/setting/repository.go (hunk 1 above)
Comment (delvh) [reply modules/setting/repository.go]: Furthermore, wouldn't it be
```suggestion
			RetargetChildrenOnMerge                    bool
```
?

File: routers/web/repo/pull.go (hunk 3 above)
Comment (kvaster) [reply routers/web/repo/pull.go]: Good idea - will do it.

**Comment by kvaster** [comment]
Actually we're using this patch already for long time, but it was difficult for me to write tests... However I can do necessary changes to PR itself...

File: routers/web/repo/pull.go (hunk 3 above)
Comment (kvaster) [reply routers/web/repo/pull.go]: I've made error handling a little bit better, but it's not possible to move the whole code to `repo_service.DeleteBranch`.

File: docs/content/doc/administration/config-cheat-sheet.en-us.md [hunk 4]
```diff
@@ -137,6 +137,7 @@ In addition there is _`StaticRootPath`_ which can be set as a built-in at build
 - `POPULATE_SQUASH_COMMENT_WITH_COMMIT_MESSAGES`: **false**: In default squash-merge messages include the commit message of all commits comprising the pull request.
 - `ADD_CO_COMMITTER_TRAILERS`: **true**: Add co-authored-by and co-committed-by trailers to merge commit messages if committer does not match author.
 - `TEST_CONFLICTING_PATCHES_WITH_GIT_APPLY`: **false**: PR patches are tested using a three-way merge method to discover if there are conflicts. If this setting is set to **true**, conflicting patches will be retested using `git apply` - This was the previous behaviour in 1.18 (and earlier) but is somewhat inefficient. Please report if you find that this setting is required.
+- `RETARGET_CHILDREN_ON_MERGE`: **false**: Retarget the child pull request to the parent pull request branch target.
```
Comment (zeripath) [inline docs/content/doc/administration/config-cheat-sheet.en-us.md]:
```suggestion
- `RETARGET_CHILDREN_ON_MERGE`: **false**: Retarget child pull requests to the parent pull request branch target on merge of parent pull request.
```

File: routers/api/v1/repo/pull.go [hunk 5]
```diff
@@ -904,6 +904,11 @@ func MergePullRequest(ctx *context.APIContext) {
 			}
 			defer headRepo.Close()
 		}
+		if err := pull_service.RetargetChildrenOnMerge(ctx, ctx.Doer, pr); err != nil {
+			log.Error("RetargetChildrenOnMerge: %v", err)
```
Comment (zeripath) [inline routers/api/v1/repo/pull.go]: This is unnecessary.

```suggestion
```

File: routers/web/repo/pull.go [hunk 6]
```diff
@@ -1399,6 +1399,13 @@ func CleanUpPullRequest(ctx *context.Context) {
 
 func deleteBranch(ctx *context.Context, pr *issues_model.PullRequest, gitRepo *git.Repository) {
 	fullBranchName := pr.HeadRepo.FullName() + ":" + pr.HeadBranch
+
+	if err := pull_service.RetargetChildrenOnMerge(ctx, ctx.Doer, pr); err != nil {
+		log.Error("RetargetChildrenOnMerge: %v", err)
```
Comment (zeripath) [inline routers/web/repo/pull.go]:
```suggestion
```

File: custom/conf/app.example.ini [hunk 7]
```diff
@@ -1062,6 +1062,9 @@ ROUTER = console
 ;;
 ;; In addition to testing patches using the three-way merge method, re-test conflicting patches with git apply
 ;TEST_CONFLICTING_PATCHES_WITH_GIT_APPLY = false
+;;
+;; Retarget the child pull request to the parent pull request branch target
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
;; Retarget child pull request to the parent pull request branch target when the parent is merged
```

**Comment by zeripath** [comment]
Thanks for this PR - sorry it's taking so long to get in.

A few thoughts:

1. Could you explain why this retargeting only occurs if the new target is in the same repo? Is this to copy Github's behaviour?
2. I think it might be better to make this on by default.
3. I suspect we would be better off making this setting part of `PullRequestsConfig` and allowing repos to set this on/off by themselves.

**Comment by kvaster** [comment]
Telling the truth, I was not aware about retargeting to another repository, but now I think it might be also usefull.
I also think it will not hurt anybody if it will be on by default. And an option inside repository itself will be even better.

**Comment by kvaster** [comment]
It was started something long ago. I think I've not prepare code for retargeting to new repository cause `ChangeTargetBranch` function was already implemented while there was no `ChangeTargetBranchAndRepo`...

**Comment by denyskon** [comment]
Closing in favor of #28539. If there is a reason to, feel free to reopen.

**Pull request closed without merging**